### PR TITLE
Use EGL over GLX

### PR DESCRIPTION
**Pull request by sfan5**
only affects the legacy CIrrDeviceLinux.
GLX is quite old and EGL is the "new" standard for this.
this change could fix some problems people have been having.

## To do

This PR is technically ready.

Note: this should be tested on Linux/Intel, Linux/AMD and Linux/NVIDIA before merge!

### CODE REVIEW

File: irr/src/CMakeLists.txt [hunk 1]
```diff
@@ -213,11 +213,14 @@ if(ENABLE_GLES2)
 	find_package(OpenGLES2 REQUIRED)
 endif()
 if(ENABLE_OPENGL)
-	find_package(OpenGL REQUIRED)
-	# we need EGL so use the import target
-	if(NOT DEFINED EGL_LIBRARY)
+	if(DEVICE STREQUAL "X11")
```
Comment (okias) [inline irr/src/CMakeLists.txt:216]: are u sure? This doesn't look right. Shouldn't be related to X11 only

**Comment by okias** [comment]
No-one should use GLX,.. I just had talk about GLX/EGL subject on XDC24 few days ago.

I assume the SDL2 revert is not meant to remain at time of merge?

File: irr/src/CMakeLists.txt (hunk 1 above)
Comment (sfan5) [reply irr/src/CMakeLists.txt:216]: "X11" here means CIrrDeviceLinux, which is the only thing using GLX.

**Comment by sfan5** [comment]
> I assume the SDL2 revert is not meant to remain at time of merge?

When we're using SDL2 it will handle the GL context for us (and I bet it already uses EGL).
This fix is specifically for the impending release, where we will temporarily disable SDL2 again.

**Comment by lhofhansl** [comment]
I wonder if FSAA will still work with this.
On my machine that works with GLX, but not SDLx.

**Comment by lhofhansl** [comment]
I had looked briefly in the past. While it's logically not that hard to add, I found it incredibly tedious and obtuse in Irrlicht.

**Comment by sfan5** [comment]
@lhofhansl actually I think the root cause is just #13397 and it's just a weird accident that it still worked with GLX.
we could possibly introduce multisample textures to bring back FSAA. not sure if worth it.

**Comment by lhofhansl** [comment]
I don't mean to say that that should block this PR.

That said, with some larger viewing ranges the experience without FSAA is really shitty and SSAA is just too slow (and FXAA does not deal with geometry aliasing at all).

**Comment by okias** [comment]
Tested on Linux, works well for me.

**Comment by sfan5** [comment]
Adding to milestone since this is at the very least a maintenance improvement and could *potentially* fix problems.

**Comment by lhofhansl** [comment]
> could possibly introduce multisample textures to bring back FSAA. not sure if worth it

Am I the only one who uses large viewing_ranges and find the geometry aliasing effects causing moire patterns grating? It looks like a cheap, bad game from the early 2000s.
FXAA or any other pixel space anti-aliasing will not fix this. SSAA does by just rendering at a higher resolution, but it's too slow.

IMHO there is no alternative really... Well other than rendering at a lower geometry resolution at a distance, perhaps with some LOD.

**Review by SmallJoker — APPROVED** [review]
Tested on Linux non-SDL X11
Enabled settings: shadows + post processing + debanding enabled.

Intel HD, i915, OpenGL 4.6. Values of `antialiasing`:

* `none`: works
* `fsaa` (16): "works" but looks like `none`
* `ssaa` produces a black screen. No errors. [1]
* `fxaa`: antialiasing works

AMD Polaris, amdgpu, OpenGL 4.6. Values of `antialiasing`:

* `none`: works
* `fsaa` (16): "works" but looks like `none`
* `ssaa` produces a black screen. No errors. [1]
* `fxaa`: antialiasing works

[1] Disabling post-processing or shaders altogether can get it to "work", but it looks like `none`.

Furthermore (not relevant for this PR), I noticed that the Debian package `libgl1-mesa-dev` is now a placeholder for transition. `libegl-dev` is now required. `libopengl-dev` is technically optional, but CMake fails to find `libOpenGL.so.0` on its own if that's missing.

**Comment by SmallJoker** [comment]
I also found an *interesting* performance issue when using `DRI_PRIME=1 vblank_mode=0 ./minetest` after switching the display cable to the dGPU. The dGPU is used but the FPS jumps all over the place. I don't think it's relevant for this PR but leaving it here just in case.

https://github.com/user-attachments/assets/a35b6a9a-769a-4676-9d0b-231236c2853d

**Comment by lhofhansl** [comment]
> fsaa (16): "works" but looks like none

FSAA definitely works for me, on Intel integrated graphics and NVidia, and it make a huge difference.
**Note** that you won't see a big difference with smaller viewing_ranges (less than, say, 500). Beyond that you'll get pretty bad geometry (not texture) aliasing.

FSAA and SSAA both handle geometry aliasing.
FXAA and SSAA both handle texture aliasing.

(And so FSAA does not help with texture aliasing, and FXAA does not help with geometry aliasing).

SSAA of course is best, but it's also by far the slowest - unusable even with my NVidia card.

**Comment by grorp** [comment]
Now I'm getting these warnings at startup:

```
libEGL warning: egl: failed to create dri2 screen
MESA-INTEL: warning: Haswell Vulkan support is incomplete
```

**Comment by grorp** [comment]
FSAA still has an effect for me (as long as post-processing) is disabled, but font rendering is somehow worse now:

screenshot with 16x FSAA, before
screenshot 1

screenshot with 16x FSAA, after
screenshot 2

screenshot with 16x FSAA, before, this time in the mainmenu
screenshot 3

screenshot with 16x FSAA, after, mainmenu
screenshot 4

**Comment by grorp** [comment]
And now it's hanging at startup with a black window and this backtrace:

```
#0  0x00007eff5da0a87d in poll () from /lib64/libc.so.6
#1  0x00007eff5d5710b2 in _xcb_conn_wait.part.0 () from /lib64/libxcb.so.1
#2  0x00007eff5d573090 in xcb_wait_for_special_event () from /lib64/libxcb.so.1
#3  0x00007eff4e27a551 in dri3_wait_for_event_locked () from /lib64/libEGL_mesa.so.0
#4  0x00007eff4e27a75b in dri3_find_back () from /lib64/libEGL_mesa.so.0
#5  0x00007eff4e27cc91 in dri3_get_buffer.isra () from /lib64/libEGL_mesa.so.0
#6  0x00007eff4e27d028 in loader_dri3_get_buffers () from /lib64/libEGL_mesa.so.0
#7  0x00007eff2ee385a5 in dri_image_drawable_get_buffers () from /usr/lib64/dri/crocus_dri.so
#8  0x00007eff2ee38653 in dri2_allocate_textures () from /usr/lib64/dri/crocus_dri.so
#9  0x00007eff2ee3bb86 in dri_st_framebuffer_validate () from /usr/lib64/dri/crocus_dri.so
#10 0x00007eff2ef18e1c in st_framebuffer_validate () from /usr/lib64/dri/crocus_dri.so
#11 0x00007eff2ef19d22 in st_manager_validate_framebuffers () from /usr/lib64/dri/crocus_dri.so
#12 0x00007eff2f1a095b in st_update_framebuffer_state () from /usr/lib64/dri/crocus_dri.so
#13 0x00007eff2f1a5b4f in st_Clear () from /usr/lib64/dri/crocus_dri.so
#14 0x0000000000c50880 in irr::video::COpenGLDriver::clearBuffers(unsigned short, irr::video::SColor, float, unsigned char) ()
#15 0x0000000000c4fa80 in irr::video::COpenGLDriver::beginScene(unsigned short, irr::video::SColor, float, unsigned char, irr::video::SExposedVideoData const&, irr::core::rect<int>*) ()
#16 0x000000000068f583 in GUIEngine::run() ()
#17 0x0000000000690160 in GUIEngine::GUIEngine(JoystickController*, irr::gui::IGUIElement*, RenderingEngine*, IMenuManager*, MainMenuData*, bool&) ()
#18 0x00000000004ff0bf in ClientLauncher::main_menu(MainMenuData*) ()
#19 0x000000000050086b in ClientLauncher::launch_game(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, bool, GameStartData&, Settings const&) ()
#20 0x000000000050335f in ClientLauncher::run(GameStartData&, Settings const&) ()
#21 0x00000000004cfe58 in main ()

```

Commit d52e4cdbd, and it's fixed when I revert aa273119f292e52df8d99f663df5337bac31556d

It only happens if I start it via the desktop file after `make install`, starting it from terminal works as expected

**Comment by lhofhansl** [comment]
I get an immediate Segmentation fault now with the NVidia driver (not Intel)

**Comment by lhofhansl** [comment]
```
2024-10-22 17:03:01: ERROR[Main]: Irrlicht: Could not create EGL surface.

Thread 1 "minetest" received signal SIGSEGV, Segmentation fault.
0x0000000000c60471 in irr::video::COpenGLDriver::~COpenGLDriver() ()
(gdb) bt
#0  0x0000000000c60471 in irr::video::COpenGLDriver::~COpenGLDriver() ()
#1  0x0000000000c60723 in virtual thunk to irr::video::COpenGLDriver::~COpenGLDriver() ()
#2  0x0000000000c6ce03 in irr::video::createOpenGLDriver(irr::SIrrlichtCreationParameters const&, irr::io::IFileSystem*, irr::video::IContextManager*) ()
#3  0x0000000000c0165f in irr::CIrrDeviceLinux::createDriver() ()
#4  0x0000000000c0c6fe in irr::CIrrDeviceLinux::CIrrDeviceLinux(irr::SIrrlichtCreationParameters const&) ()
#5  0x0000000000bfff30 in createDeviceEx ()
#6  0x00000000005f9e12 in RenderingEngine::RenderingEngine(MyEventReceiver*) ()
#7  0x00000000004fea74 in ClientLauncher::init_engine() ()
#8  0x0000000000503d0f in ClientLauncher::run(GameStartData&, Settings const&) ()
#9  0x00000000004d13d8 in main ()
```

**Comment by sfan5** [comment]
Ok then let's not invest more time into legacy shit.

**Comment by okias** [comment]
For the xcb, looks very similar to https://gitlab.freedesktop.org/xorg/lib/libxcb/-/issues/38 .

Going back to GLX seriously doesn't make sense, just workarounding buggy X11 should be enough

**Pull request merged**